# Pima Data cleaning



In [3]:
# Import Modules Here
import pandas as pd 

## Part 1: Read the Data into Jupyter

In [5]:
# Read the data into a dataframe named pima_df.
pima_df = pd.read_csv('/home/mariahwarren/diabetes_uncleaned_data.csv')
pima_df

Pregnancies  Glucose BloodPressure  SkinThickness  Insulin   BMI  \
0             3.0    167.0            39             28      545  53.0   
1            13.0    151.0           107             38    10000   9.2   
2             8.0     59.0           109             18      739  59.1   
3            11.0    132.0            83             11      376   4.1   
4            10.0    165.0            48             37      737  39.4   
...           ...      ...           ...            ...      ...   ...   
999           6.0    110.0            26             27      590  20.8   
1000          6.0      NaN            10              4      640  55.7   
1001          4.0     80.0            24              6      268  48.9   
1002          3.0    142.0         Error             10      299   4.8   
1003          6.0    151.0            60             26      362  43.4   

      DiabetesPedigreeFunction    Age Outcome  
0                        2.447  125.0   FALSE  
1                        0.507  148.0   FALSE  
2                        1.769   91.0    TRUE  
3                        2.772    NaN    TRUE  
4                        0.729   73.0   FALSE  
...                        ...    ...     ...  
999                      1.185   95.0   FALSE  
1000                     2.483   79.0   FALSE  
1001                     0.989   81.0    TRUE  
1002                     2.411   71.0    TRUE  
1003                     2.027   23.0   FALSE  

[1004 rows x 9 columns]

## Part 2: Missing Data

In [7]:
#Question 1:  Find Columns with missing data
pima_df.isna().sum()

Pregnancies                 114
Glucose                      16
BloodPressure                22
SkinThickness                 0
Insulin                       0
BMI                           1
DiabetesPedigreeFunction      0
Age                          22
Outcome                       0
dtype: int64

In [9]:
# Question 2:  Write code to remove rows with missing data and save the new dataframe as pima_missing_fixed_df
pima_missing_fixed_df = pima_df.dropna(how='any')
pima_missing_fixed_df

Pregnancies  Glucose BloodPressure  SkinThickness  Insulin   BMI  \
0             3.0    167.0            39             28      545  53.0   
1            13.0    151.0           107             38    10000   9.2   
2             8.0     59.0           109             18      739  59.1   
4            10.0    165.0            48             37      737  39.4   
5          -100.0    142.0            88             39       82  42.7   
...           ...      ...           ...            ...      ...   ...   
998           6.0    110.0            26             27      590  20.8   
999           6.0    110.0            26             27      590  20.8   
1001          4.0     80.0            24              6      268  48.9   
1002          3.0    142.0         Error             10      299   4.8   
1003          6.0    151.0            60             26      362  43.4   

      DiabetesPedigreeFunction    Age Outcome  
0                        2.447  125.0   FALSE  
1                        0.507  148.0   FALSE  
2                        1.769   91.0    TRUE  
4                        0.729   73.0   FALSE  
5                        1.369  148.0    TRUE  
...                        ...    ...     ...  
998                      1.185   95.0   FALSE  
999                      1.185   95.0   FALSE  
1001                     0.989   81.0    TRUE  
1002                     2.411   71.0    TRUE  
1003                     2.027   23.0   FALSE  

[840 rows x 9 columns]

In [11]:
# Question 3:  Print out the number of rows in the dataframe.
print('Total Rows:', len(pima_missing_fixed_df.index))

Total Rows: 840


### Question 4:  Discuss why missing data could be a problem.  What other methods could you use in this situation?

Missing data could be a problem because it affects your analysis of the data and what conclusions you can draw from it. Other methods I could use are fillna(), fillna().mean(), interpolate(), and scikit-learn.

## Part 3: Duplicated Data

In [14]:
# Question 1: find the rows of duplicated data in the
# pima_missing_fixed_df and print them.
pima_missing_fixed_df.duplicated().sum()

4

In [16]:
# Question 2: Write code to remove duplicate rows
# and save the new dataframe as pima_missing_fixed_df
pima_missing_fixed_df = pima_missing_fixed_df.drop_duplicates()

In [18]:
# Question 3:  Print out the number of rows in the dataframe.
print('Total Rows:', len(pima_missing_fixed_df.index))

Total Rows: 836


### Question 4:  In your jupyter notebook, discuss why duplicate data could be a problem.
### What other methods could you use in this situation?

Duplicate data could be a problem because it affects your analysis of the data and what conclusions you can draw from it. Other methods I could use are duplicated().

## Part 4: Mis-typed Columns/Broken Data


In [34]:
# Question 1: find the columns where most of the data seems to be one type of data, and there seems to be a data error. In the markdown, after exploring in code, add a markdown cell discussing which columns have the data error.
pima_missing_fixed_df['Pregnancies'].value_counts()
pima_missing_fixed_df['Glucose'].value_counts()
pima_missing_fixed_df['BloodPressure'].value_counts()
pima_missing_fixed_df['SkinThickness'].value_counts()
pima_missing_fixed_df['Insulin'].value_counts()
pima_missing_fixed_df['BMI'].value_counts()
pima_missing_fixed_df['DiabetesPedigreeFunction'].value_counts()
pima_missing_fixed_df['Age'].value_counts()
pima_missing_fixed_df['Outcome'].value_counts()

Outcome
FALSE    441
TRUE     394
ERROR      1
Name: count, dtype: int64

In [56]:
pima_missing_fixed_df[pima_missing_fixed_df['Pregnancies']=='ERROR']
pima_missing_fixed_df[pima_missing_fixed_df['Glucose']=='ERROR']
pima_missing_fixed_df[pima_missing_fixed_df['BloodPressure']=='ERROR']
pima_missing_fixed_df[pima_missing_fixed_df['SkinThickness']=='ERROR']
pima_missing_fixed_df[pima_missing_fixed_df['Insulin']=='ERROR']
pima_missing_fixed_df[pima_missing_fixed_df['BMI']=='ERROR']
pima_missing_fixed_df[pima_missing_fixed_df['DiabetesPedigreeFunction']=='ERROR']
pima_missing_fixed_df[pima_missing_fixed_df['Age']=='ERROR']
pima_missing_fixed_df[pima_missing_fixed_df['Outcome']=='ERROR']

Pregnancies  Glucose BloodPressure  SkinThickness  Insulin   BMI  \
759          2.0     69.0            61             25      393  36.1   

     DiabetesPedigreeFunction    Age Outcome  
759                     0.026  141.0   ERROR

The Outcome column has the data error in row 759.

In [59]:
# Question 2: Remove the rows with data errors
pima_missing_fixed_df = pima_missing_fixed_df.loc[pima_missing_fixed_df['Outcome']!= 'ERROR', :]

In [63]:
# Question 3: After the data is fixed in your columns change the columns to the correct type and save this as the pima_fixed_columns_df.
def word_bool(x: str):
    if x.lower() == 'false':
        return False
    elif x.lower() == 'true':
        return True
    else:
        raise Exception('this must be true or false case insensitive')


pima_fixed_columns_df = pima_missing_fixed_df.copy()
pima_fixed_columns_df['Outcome'] = pima_fixed_columns_df['Outcome'].apply(word_bool)
print(pima_fixed_columns_df.dtypes)
pima_fixed_columns_df['Outcome'].value_counts()

Pregnancies                 float64
Glucose                     float64
BloodPressure                object
SkinThickness                 int64
Insulin                       int64
BMI                          object
DiabetesPedigreeFunction    float64
Age                         float64
Outcome                        bool
dtype: object


Outcome
False    441
True     394
Name: count, dtype: int64

In [65]:
# Question 4: Run pima_fixed_columns_df.info() to confirm the columns have changed.
pima_fixed_columns_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 835 entries, 0 to 1002
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               835 non-null    float64
 1   Glucose                   835 non-null    float64
 2   BloodPressure             835 non-null    object 
 3   SkinThickness             835 non-null    int64  
 4   Insulin                   835 non-null    int64  
 5   BMI                       835 non-null    object 
 6   DiabetesPedigreeFunction  835 non-null    float64
 7   Age                       835 non-null    float64
 8   Outcome                   835 non-null    bool   
dtypes: bool(1), float64(4), int64(2), object(2)
memory usage: 59.5+ KB


## Part 5: Outlier Detection and Removal


In [91]:
# Question 1: Print out the Outliers in each column in the pima_fixed_columns_df dataframe, use the IQR method of outlier detection.
def is_outlier(column: pd.Series):
    Q1 = column.quantile(0.25)
    Q3 = column.quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    return ~column.between(lower_bound, upper_bound)


display(pima_fixed_columns_df[is_outlier(pima_fixed_columns_df['Pregnancies'])])
display(pima_fixed_columns_df[is_outlier(pima_fixed_columns_df['Glucose'])])
display(pima_fixed_columns_df[is_outlier(pima_fixed_columns_df['SkinThickness'])])
display(pima_fixed_columns_df[is_outlier(pima_fixed_columns_df['Insulin'])])
display(pima_fixed_columns_df[is_outlier(pima_fixed_columns_df['DiabetesPedigreeFunction'])])
display(pima_fixed_columns_df[is_outlier(pima_fixed_columns_df['Age'])])

Pregnancies  Glucose BloodPressure  SkinThickness  Insulin   BMI  \
5       -100.0    142.0            88             39       82  42.7   

   DiabetesPedigreeFunction    Age  Outcome  
5                     1.369  148.0     True

Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

Pregnancies  Glucose BloodPressure  SkinThickness  Insulin  BMI  \
1         13.0    151.0           107             38    10000  9.2   

   DiabetesPedigreeFunction    Age  Outcome  
1                     0.507  148.0    False

Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

In [113]:
# Question 2: Use loc to remove outliers in each of the columns that have outliers, save this as pima_outlier_removed_df.
pima_outlier_removed_df = pima_fixed_columns_df[is_outlier(pima_fixed_columns_df['Pregnancies']) != True]
pima_outlier_removed_df = pima_fixed_columns_df[is_outlier(pima_fixed_columns_df['Insulin']) != True]
pima_outlier_removed_df

Pregnancies  Glucose BloodPressure  SkinThickness  Insulin   BMI  \
0             3.0    167.0            39             28      545  53.0   
2             8.0     59.0           109             18      739  59.1   
4            10.0    165.0            48             37      737  39.4   
5          -100.0    142.0            88             39       82  42.7   
6             6.0    151.0            60             26      362  43.4   
...           ...      ...           ...            ...      ...   ...   
996           6.0    117.0            91             28      212   1.0   
997          13.0    163.0            10             37      457   8.0   
998           6.0    110.0            26             27      590  20.8   
1001          4.0     80.0            24              6      268  48.9   
1002          3.0    142.0         Error             10      299   4.8   

      DiabetesPedigreeFunction    Age  Outcome  
0                        2.447  125.0    False  
2                        1.769   91.0     True  
4                        0.729   73.0    False  
5                        1.369  148.0     True  
6                        2.027   23.0    False  
...                        ...    ...      ...  
996                      1.969  137.0    False  
997                      2.400  125.0     True  
998                      1.185   95.0    False  
1001                     0.989   81.0     True  
1002                     2.411   71.0     True  

[834 rows x 9 columns]

In [115]:
# Question 3:  Print out the row count in the pima_outlier_removed_df and confirm this number is correct.
print('Total Rows:', len(pima_outlier_removed_df.index))

Total Rows: 834
